src/pages/SearchPage/SearchPage.test.js

```diff
@@ -0,0 +1,210 @@
+    const { getByTestId, findByText } = setup();
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    await findByText('0 results returned.');
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Will show an error message when the API call fails', async () => {
+    const { getByTestId, findByText } = setup('fail');
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    await findByText('There was an error processing your request.');
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Will load heatmap with real test data', async () => {
+    const { getByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(getByTestId('heatmap')).toBeInTheDocument();
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+});
+
+describe('Heatmap values', () => {
+  test('Heatmap values matches dummy data', async () => {
+    const dummyHeatmapValues = getHeatmapValues(dummyPosts);
+    const { getByTestId, getAllByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    // Checking buttons
+    const heatmapValues = getAllByTestId('heatmap-button').map((button) => parseInt(button.textContent, 10));
+    // Make sure number of buttons in heatmap is correct
+    expect(heatmapValues.length).toEqual(NUM_GRID_ITEMS);
+    // Heatmap values must match dummy data values
+    expect(heatmapValues).toEqual(dummyHeatmapValues);
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+  test('Heatmap elements use the correct theme colours', async () => {
+    const { getByTestId, getAllByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    // Checking buttons
+    const heatmapButtons = getAllByTestId('heatmap-button');
+    for (let i = 0; i < heatmapButtons.length; i += 1) {
+      const button = heatmapButtons[i];
+      const buttonValue = parseInt(button.textContent, 10);
+      if (buttonValue <= 10) {
+        expect(button).toHaveStyleRule('background', theme.heatmapColors[buttonValue]);
+      } else {
+        expect(button).toHaveStyleRule('background', theme.heatmapColors[10]);
+      }
+    }
+  });
+  test('Heatmap displays the correct timezone message', async () => {
+    const { getByTestId } = setup('pass', dummyPosts);
+    await waitForElementToBeRemoved(getByTestId('loading-spinner'));
+    expect(getByTestId('heatmap')).toBeInTheDocument();
+    // Time message is showing
+    const timeMessage = getByTestId('time-message');
+    expect(timeMessage.textContent).toBe('All times are shown in your timezone: UTC');
+    expect(axiosMock.get).toHaveBeenCalledTimes(1);
+  });
+});
+
+function getPosts(data, day, hour) {
+  const currentTable = data[day][hour];
+  const posts = [];
+  for (let i = 0; i < currentTable.length; i += 1) {
```

Comment:
You could use `.map` here

```
const posts = currentTable.map(({ author, title }) => ({
  ...
)};
```

Reply:
Trying to `map` it but I'm not sure if it's possible with what I'm doing. I'm creating a 1d array from all of the table cells so I can test each one easily. The reason for doing this was because it became to complex to loop and test the data while it was in its complex object form. This is why there are multiple `push` statements in each loop:

```js
    posts.push(author);
    posts.push(title);
    posts.push(created);
    posts.push(score);
    posts.push(num_comments);
```

`currentTable`'s values are rows in the table which are 5 different key/value pairs. `map` is only going to give me one opportunity/return to deal with all of the key/values for that particular row so I would only be able to return an entire row instead of returning each value so that the data is in a flattened 1d array.

The only thing I can think of at the moment is to keep it as a row (2d array) from the `.map` array and then use `.flat` to flatten it before it's returned.

Reply:
I think I might be able to do it with 2 maps, the 2nd looping through keys. Attempting this at the moment and then will do a comparison at the end.

Reply:
Btw the name `posts` is a bit confusing here. That's why I expected it to be an array of post objects.

Reply:
> 
> 
> Btw the name `posts` is a bit confusing here. That's why I expected it to be an array of post objects.

You're right. I changed it as soon as you did the review as I immediately saw it was the cause of confusion. It was originally for each post, but the code evolved quite a bit and naming got lost along the way. Thanks.

Reply:
I don't think it's going to be possible to do it elegantly with array methods, namely for the following reason:

1. I can't map through all of the keys as I have keys in my object, such as the Reddit post URL, but I'm not testing for the Reddit URL in the test.

2. The test itself loops through the array and checks for the item's text to be in the document. Since the data is a 1d array with no keys/values, I can't separate out content, such as the URL, so that it's not being tested for. I can't switch back to passing an object to the test as that creates massive problems when it comes to testing and trying to loop through each of the objects.

What I've got is, but the test will break for reasons above. I have to flatten the solution below otherwise I have nested arrays:

```js
function getPosts(data, day, hour) {
  const currentTable = data[day][hour];
  const cells = currentTable.map((row) => {
    return Object.keys(row).map(key => {
      if (key === 'created') {
        return displayHHMM(row[key]);
      }
      return row[key].toString();
    });
  }).flat();
  return cells;
}
```

I hate giving up on a coding problem, but I'm starting to think that the way I did initially is going to be one of the more elegant solutions to handle it. 

I await your input. I don't mind going at it for as long as I need to get it right, just not sure if I'm going to find something that works better.

Reply:
Sorry, the code in the comment was wrong. I wanted to use `reduce` instead of `map`.

```
const posts = currentTable.reduce((tmpPosts, row) => ({
      // Convert date obj to "hh:mm" with am/pm suffixed
    const created = displayHHMM(row.created);
    // Can't test numeric values, need to convert to string
    const num_comments = row.num_comments.toString();
    const score = row.score.toString();
  return tmpPosts.concat(row.author, row.title, created, score, num_comments);
)}, []);
```

Reply:
This will give you a 1D array with exactly like the one in your code if I'm right

Reply:
Amazing!! I tried it the first time (I knew to switch `map` with `reduce`) but couldn't figure out the syntax error it had.

Here it is in its glory. Thank you so much!

```js
function getCellData(data, day, hour) {
  const currentTable = data[day][hour];
  const cells = currentTable.reduce((tmpCells, row) => {
    // Convert date obj to "hh:mm" with am/pm suffixed
    const created = displayHHMM(row.created);
    // Can't test numeric values, need to convert to string
    const num_comments = row.num_comments.toString();
    const score = row.score.toString();
    return tmpCells.concat(row.author, row.title, created, score, num_comments);
  }, []);
  return cells;
}
```